### PR TITLE
[FSDP][state_dict] Avoid assigning the root _device_mesh to the children _device_mesh

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114384

Assigning the root _device_mesh to the children _device_mesh is not correct as each FSDP state can have a different DeviceMesh. We are also replacing fully_shard with a new implementation. So there is no need to worry about the fully_shard behavior.

Differential Revision: [D51507959](https://our.internmc.facebook.com/intern/diff/D51507959/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

### CODE REVIEW

**Comment by XilunWu** [comment]
> as each FSDP state can have a different DeviceMesh.

Does this mean `fsdp_state` should not be associated with a `DeviceMesh` object because it doesn't necessarily attach to the root device mesh?

**Comment by fegin** [comment]
> > as each FSDP state can have a different DeviceMesh.
> 
> Does this mean `fsdp_state` should not be associated with a `DeviceMesh` object because it doesn't necessarily attach to the root device mesh? Therefore it's no longer meaningful to keep the `_device_mesh` attribute, is that right?

No. This means that `fsdp_state` should have its own `DeviceMesh`. The complexity comes from the original fully_shard only has a different way to maintain the state from the FSDP does.

**Review by wz337 — COMMENTED** [review]
LGTM. But I feel like some existing tests will fail, adding the periodic tag to surface those tests that need to be changed for this change.

**Review by wz337 — APPROVED** [review]
LGTM

**Review by wz337 — APPROVED** [review]

**Comment by wz337** [comment]
@wz337 Remind myself to add a unit test for this use case.

File: torch/distributed/fsdp/_optim_utils.py [hunk 1]
```diff
@@ -2063,13 +2063,13 @@ def return_fn(fqn_to_param_info):
 
 @no_type_check
 def _set_optim_use_dtensor(
-    module: nn.Module,
+    fsdp_state: _FSDPState,
```
Comment (wz337) [inline torch/distributed/fsdp/_optim_utils.py:2066]: I think we also need to update the caller of `_set_optim_use_dtensor` to pass fsdp_state into `_set_optim_use_dtensor` as well.

https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/fully_sharded_data_parallel.py#L767C27-L767C27

Comment (fegin) [reply torch/distributed/fsdp/_optim_utils.py:2066]: For a FullyShardedDataParallel module, itself is the fsdp_state.

**Pull request closed without merging**